### PR TITLE
JCLOUDS-1460: Feat/adding aws instances

### PR DESCRIPTION
**Pull request by trentontrees**
PR #1244 also does this feature.  Slightly different approach to the problem.
[Jira 1460](https://issues.apache.org/jira/browse/JCLOUDS-1460)

Added support for:

t2.nano
t2.xlarge
t2.2xlarge
t3.nano
t3.micro
t3.small
t3.medium
t3.large
t3.xlarge
t3.2xlarge
m4.16xlarge
AWS Instance Type reference page:
https://aws.amazon.com/ec2/instance-types/

### CODE REVIEW

**Comment by trentontrees** [comment]
Looking at the build failure, I'm not understanding how my PR affected the Nova API.  Mine changes should be local to AWS only.

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 1]
```diff
@@ -429,6 +444,89 @@ public static EC2HardwareBuilder t2_large() {
...
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
+   }
+
+   /**
+    * @see InstanceType#T3_LARGE
+    */
+   public static EC2HardwareBuilder t3_large() {
+      return new EC2HardwareBuilder(InstanceType.T3_LARGE).t3()
+              .ram(8192)
+              .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (danielestevez) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java]: T3 instances has number of vCPU is wrong https://aws.amazon.com/it/ec2/instance-types/t3/
The javadoc is ok at ```InstanceType```, though

Could you please review them?

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java [hunk 2]
```diff
@@ -429,6 +444,89 @@ public static EC2HardwareBuilder t2_large() {
             .processors(ImmutableList.of(new Processor(1.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
    }
 
+   /**
+    * @see InstanceType#T2_XLARGE
+    */
+   public static EC2HardwareBuilder t2_xlarge() {
+      return new EC2HardwareBuilder(InstanceType.T2_XLARGE).t2()
+              .ram(16384)
+              .processors(ImmutableList.of(new Processor(2.0, 0.4))).rootDeviceType(RootDeviceType.EBS);
```
Comment (danielestevez) [inline apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java]: Also t2 number of processors seem to be wrong, t2.xlarge has 4vCPU

**Comment by danielestevez** [comment]
rebuild please

File: apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java (hunk 2 above)
Comment (trentontrees) [reply apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java]: https://aws.amazon.com/it/ec2/instance-types/t2/

I wasn't 100% sure on my values, but i used the existing `t2.large` and `t2.xlarge` as my example space.  From the above link I'm seeing `t2.large(vCpu=2)` and `t2.xlarge(vCpu=4)`

Should I change all of the t2 processors to match the product detail page?

Comment (trentontrees) [reply apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java]: After looking at the M4 for additional sanity, I agree with what you are proposing and will make the change soon.

Comment (danielestevez) [reply apis/ec2/src/main/java/org/jclouds/ec2/compute/domain/EC2HardwareBuilder.java]: Code should match what the AWS documentation and the javadoc, yes.

t2.large is ok but t2.xlarge should be 4 yes

**Comment by trentontrees** [comment]
@danielestevez This PR should be set for review (again)

**Comment by danielestevez** [comment]
That link looks useful, thanks!
Please squash the commits and we're done

**Comment by trentontrees** [comment]
Took a few tries, but successfully squashed the PR.  Should be good to go :)

Out of curiosity, why have the original repository squash the commits?  Why not use github's squash and merge option to flatten the commits?  I only ask as I made a squash mistake and was lucky enough to have had a few copies of my code around to quickly fix the introduced bug.

**Comment by nacx** [comment]
Thanks for the effort! Basically, because we don't push code to GitHub but to the Apache Git repository. This is just a mirror of that one and whatever is pushed just here gets removed when the Apache Git repo is mirrored.

Now that you have gotten into the wonderful world of rebasing :) would you mind amending the commit message and including the JIRA issue ID in it? This way the commit will appear in JIRA too.

**Comment by danielestevez** [comment]
Pushed to [master](http://git-wip-us.apache.org/repos/asf/jclouds/commit/219e2958) and [2.1.x](http://git-wip-us.apache.org/repos/asf/jclouds/commit/e4c99a07)! 
Thanks @trentontrees for the contribution!!

**Pull request closed without merging**